File: cloud_testing/cloud_api_poc/src/main/java/cloud_api_poc/MatrixCallable.java
package cloud_api_poc;

import static cloud_api_poc.Utils.fatalError;

import com.google.testing.Testing.Projects.TestMatrices.Create;
import com.google.testing.model.TestMatrix;
import java.util.List;
import java.util.concurrent.Callable;

class MatrixCallable implements Callable {
  private List<String> testMatrixIds;
  private Create testMatrixCreate;

  MatrixCallable(List<String> testMatrixIds, Create testMatrixCreate) {
    this.testMatrixIds = testMatrixIds;
    this.testMatrixCreate = testMatrixCreate;
  }

  private void run() {
    TestMatrix testMatrix = null;
    try {
      testMatrix = testMatrixCreate.execute();
    } catch (Exception e) {
      try {
        testMatrix = testMatrixCreate.execute();
      } catch (Exception e2) {
        fatalError(e2);
      }
    }

    if (testMatrix == null) {
      throw new IllegalStateException("test matrix is null");
    }

    testMatrixIds.add(testMatrix.getTestMatrixId());
  }

  @Override
  public Object call() throws Exception {
    run();
    return null;
  }
}

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/MatrixCallable.kt
package ftl

import com.google.testing.Testing.Projects.TestMatrices.Create
import com.google.testing.model.TestMatrix
import java.util.concurrent.Callable

import ftl.Utils.fatalError
import ftl.Utils.sleep

internal class MatrixCallable(private val testMatrixIds: MutableList<String>,
                              private val testMatrixCreate: Create) : Callable<Any> {

    /** Returns test matrix id  */
    private fun executeTestMatrix(): String {
        var testMatrix: TestMatrix? = null
        try {
            testMatrix = testMatrixCreate.execute()
        } catch (e: Exception) {
            try {
                testMatrix = testMatrixCreate.execute()
            } catch (e2: Exception) {
                try {
                    testMatrix = testMatrixCreate.execute()
                } catch (e3: Exception) {
                    fatalError(e3)
                }

            }

        }

        if (testMatrix == null) {
            throw IllegalStateException("test matrix is null")
        }

        return testMatrix.testMatrixId
    }

    private fun run() {
        val tryCount = 3

        var testMatrixId: String? = null
        var infrastructureFailure = false
        for (i in 0..tryCount - 1) {
            testMatrixId = executeTestMatrix()

            while (TestMatrixState.validatingOrPending(testMatrixId)) {
                sleep(20 * 1000)
            }

            if (TestMatrixState.infastructureFailure(testMatrixId)) {
                println("Retrying infrastructure failure")
                infrastructureFailure = true
            } else {
                infrastructureFailure = false
                break
            }
        }

        if (infrastructureFailure) {
            throw RuntimeException(
                    "Infrastructure failure on test matrix after 3x retry " + testMatrixId!!)
        }

        println("Successfully scheduled test matrix: " + testMatrixId!!)

        testMatrixIds.add(testMatrixId)
    }

    @Throws(Exception::class)
    override fun call() : Void {
        run()
        return Void.TYPE.newInstance()
    }
}
